ion-schema/src/model/ranges.rs

```diff
@@ -0,0 +1,173 @@
+use crate::result::IonSchemaError;
+use std::fmt::{Display, Formatter};
+use std::ops::{Bound, Range, RangeBounds, RangeFrom, RangeInclusive, RangeTo, RangeToInclusive};
+
+#[derive(Debug, PartialEq, Clone)]
+pub struct IonSchemaRange<T: PartialOrd> {
+    min: Bound<T>,
+    max: Bound<T>,
+}
+
+impl<T: PartialOrd + Display> IonSchemaRange<T> {
+    pub fn try_new(min: Bound<T>, max: Bound<T>) -> Result<Self, IonSchemaError> {
+        match (min, max) {
+            (Bound::Unbounded, Bound::Unbounded) => Err(IonSchemaError::InvalidSchemaError {
+                description: "Invalid range".to_string(),
+            }),
+            (Bound::Excluded(x), Bound::Included(y)) if x >= y => {
+                Err(IonSchemaError::InvalidSchemaError {
+                    description: "Invalid range".to_string(),
+                })
+            }
+            (Bound::Included(x), Bound::Excluded(y)) if x >= y => {
+                Err(IonSchemaError::InvalidSchemaError {
+                    description: "Invalid range".to_string(),
+                })
+            }
+            (Bound::Excluded(x), Bound::Excluded(y)) if x >= y => {
+                Err(IonSchemaError::InvalidSchemaError {
+                    description: "Invalid range".to_string(),
+                })
+            }
+            (Bound::Included(x), Bound::Included(y)) if x > y => {
+                Err(IonSchemaError::InvalidSchemaError {
+                    description: "Invalid range".to_string(),
+                })
+            }
+            (min, max) => Ok(IonSchemaRange { min, max }),
+        }
+    }
```

Comment:
This `match` is very repetitive, but also very resistant to refactoring for brevity! 😅

While I don't think the arms/guards can be consolidated, you can defer the bodies in the error case:

```suggestion
    pub fn try_new(min: Bound<T>, max: Bound<T>) -> Result<Self, IonSchemaError> {
        match (min, max) {
            (Bound::Unbounded, Bound::Unbounded) => {},
            (Bound::Excluded(x), Bound::Included(y)) if x >= y => {},
            (Bound::Included(x), Bound::Excluded(y)) if x >= y => {},
            (Bound::Excluded(x), Bound::Excluded(y)) if x >= y => {},
            (Bound::Included(x), Bound::Included(y)) if x > y => {}
            (min, max) => return Ok(IonSchemaRange { min, max }),
        };
        Err(IonSchemaError::InvalidSchemaError {
            description: "Invalid range".to_string(),
        })
    }
```